server/infinispan/src/main/java/org/hawkular/apm/server/infinispan/InfinispanPropertyCache.java

```diff
@@ -0,0 +1,186 @@
+import org.infinispan.manager.CacheContainer;
+import org.infinispan.notifications.Listener;
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryModified;
+import org.infinispan.notifications.cachelistener.event.CacheEntryModifiedEvent;
+import org.infinispan.notifications.cachelistener.event.impl.EventImpl;
+
+/**
+ * @author Pavol Loffay
+ */
+public class InfinispanPropertyCache implements PropertyService, ServiceLifecycle {
+
+    private static final Logger log = Logger.getLogger(InfinispanPropertyCache.class.getName());
+
+    protected static final String CACHE_NAME = "property";
+
+    @Resource(lookup = "java:jboss/infinispan/APM")
+    private CacheContainer cacheContainer;
+
+    private Cache<String, Property<?>> propertyCache;
+    private CacheEventListener eventListener;
+
+
+    public InfinispanPropertyCache() {
+        this.eventListener = new CacheEventListener();
+    }
+
+    public InfinispanPropertyCache(Cache<String, Property<?>> cache) {
+        this();
+        this.propertyCache = cache;
+        this.propertyCache.addListener(eventListener);
+    }
+
+    @Override
+    @PostConstruct
+    public void init() {
+        // If cache container not already provisions, then must be running outside of a JEE
+        // environment, so create a default cache container
+        if (cacheContainer == null) {
+            log.fine("Using default cache");
+            propertyCache = InfinispanCacheManager.getDefaultCache(CACHE_NAME);
+        } else {
+            log.fine("Using container provided cache");
+            propertyCache = cacheContainer.getCache(CACHE_NAME);
+        }
+
+        propertyCache.addListener(eventListener);
+    }
+
+    @Override
+    public Property get(String tenantId, String id) {
+        Property<?> property = propertyCache.get(id);
+
+        if (property == null) {
+            property = getFromSystemOrEnvVariables(id);
+            if (property == null) {
+                return null;
+            }
+
+            try {
+                store(tenantId, property);
+            } catch (CacheException e) {
+                log.severe(String.format("Could not store environmental property: %s to cache", property));
+            }
+        }
+
+        if (log.isLoggable(Level.FINEST)) {
+            log.finest("Get property [id=" + id + "] = " + property);
+        }
+
+        return property;
+    }
+
+    @Override
+    public void store(String tenantId, List<Property<?>> properties) throws CacheException {
+        if (cacheContainer != null) {
+            propertyCache.startBatch();
+        }
+
+        for (Property<?> property: properties) {
+            if (log.isLoggable(Level.FINEST)) {
+                log.finest("Store property [id=" + property.getId() + "]: " + property);
+            }
+
+            propertyCache.put(property.getId(), property);
+        }
+
+        if (cacheContainer != null) {
+            propertyCache.endBatch(true);
+        }
```

Comment:
Very minor thing, that doesn't block this PR at all: `properties.forEach(this::store)` could be used instead of these 7 lines.